dispatch/dispatch.go

```diff
@@ -0,0 +1,423 @@
+		err := d.spawnWorker()
+		if err != nil {
+			return err
+		}
+	}
+
+	atomic.SwapUint32(&d.running, 1)
+	return nil
+}
+
+// stop stops the service and shuts down all worker routines
+func (d *Dispatcher) stop() error {
+	if !atomic.CompareAndSwapUint32(&d.running, 1, 0) {
+		return errors.New(errCannotShutdown)
+	}
+	close(d.shutdown)
+	ch := make(chan struct{})
+	timer := time.NewTimer(1 * time.Second)
+	defer func() {
+		timer.Stop()
+		select {
+		case <-timer.C:
+		default:
+		}
+	}()
+	go func(ch chan struct{}) { d.wg.Wait(); ch <- struct{}{} }(ch)
+	select {
+	case <-ch:
+		// close all routes
+		for key := range d.routes {
+			for i := range d.routes[key] {
+				close(d.routes[key][i])
+			}
+
+			d.routes[key] = nil
+		}
+
+		for len(d.jobs) != 0 { // drain jobs channel for old data
+			<-d.jobs
+		}
+
+		log.Debugln(log.DispatchMgr, "Dispatch manager shutdown.")
+
+		return nil
+	case <-timer.C:
+		return errors.New(errShutdownRoutines)
+	}
+}
+
+// isRunning returns if the dispatch system is running
+func (d *Dispatcher) isRunning() bool {
+	return atomic.LoadUint32(&d.running) == 1
+}
+
+// dropWorker deallocates a worker routine
+func (d *Dispatcher) dropWorker() error {
+	oldC := atomic.LoadInt64(&d.count)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	d.shutdown <- &wg
+	wg.Wait()
+	newC := atomic.LoadInt64(&d.count)
+
+	if oldC == newC {
+		return errors.New("dispatcher worker counts are off")
+	}
+	return nil
+}
+
+// spawnWorker allocates a new worker for job processing
+func (d *Dispatcher) spawnWorker() error {
+	if atomic.LoadInt64(&d.count) >= d.maxWorkers {
+		return errors.New("dispatcher cannot spawn more workers; ceiling reached")
+	}
+	var spawnWg sync.WaitGroup
+	spawnWg.Add(1)
+	go d.relayer(&spawnWg)
+	spawnWg.Wait()
+	return nil
+}
+
+// Relayer routine relays communications across the defined routes
+func (d *Dispatcher) relayer(i *sync.WaitGroup) {
+	atomic.AddInt64(&d.count, 1)
+	d.wg.Add(1)
+	timeout := time.NewTimer(0)
+	i.Done()
+	for {
+		select {
+		case j := <-d.jobs:
+			d.rMtx.RLock()
+			if _, ok := d.routes[j.ID]; !ok {
+				d.rMtx.RUnlock()
+				continue
+			}
+			// Channel handshake timeout feature if a channel is blocked for any
+			// period of time due to an issue with the receiving routine.
+			// This will wait on channel then fall over to the next route when
+			// the timer actuates and continue over the route list. Have to
+			// iterate across full length of routes so every routine can get
+			// their new info, cannot be buffered as we dont want to have an old
+			// orderbook etc contained in a buffered channel when a routine
+			// actually is ready for a receive.
+			// TODO: Need to consider optimal timer length
+			for i := range d.routes[j.ID] {
+				if !timeout.Stop() { // Stop timer before reset
+					// Drain channel if timer has already actuated
+					select {
+					case <-timeout.C:
+					default:
+					}
+				}
+
+				timeout.Reset(DefaultHandshakeTimeout)
```

Comment:
This is a cool solution! My concern around it is that it requires all receivers to comply with this timeout or else they miss updates, but that's way better than an infinitely increasing queue!

Increasing this time will effect every receiver of all types, so anything that does want to process every update dispatched (for example, to save each update to a database) must create their own overhead to process them. Since the dispatcher is shared, you can't have timeouts defined per receiver, but we could define a timeout per type of operation later.

This isn't a nit, but I'm curious about how it will go in future when we develop more uses for this

Reply:
Yeah, it was a hard design choice, making sure that every routine had a receiver which in effect knocked on and created a massive slow down of the dispatch system and workers. I guess if we had a massive pool of workers it wouldn't be such an issue. But if a routine isn't there within the time it will just continue to the next routine that needs to have that data. I guess we can play around with design and have a dedicated channel with no timeout for critical subsystem information.